internal/pkg/policy/policy_output.go

```diff
@@ -137,6 +192,119 @@ func (p *PolicyOutput) Prepare(ctx context.Context, zlog zerolog.Logger, bulker
 	return nil
 }
 
+func fetchAPIKeyRoles(ctx context.Context, b bulk.Bulk, apiKeyID string) (*RoleT, error) {
+	res, err := b.APIKeyRead(ctx, apiKeyID, true)
+	if err != nil {
+		return nil, err
+	}
+
+	roleMap, err := smap.Parse(res.RoleDescriptors)
+	if err != nil {
+		return nil, err
+	}
+	r := &RoleT{
+		Raw: res.RoleDescriptors,
+	}
+
+	// Stable hash on permissions payload
+	if r.Sha2, err = roleMap.Hash(); err != nil {
+		return nil, err
+	}
+
+	return r, nil
+}
+
+// mergeRoles takes old and new role sets and merges them following these rules:
+// - take all new roles
+// - append all old roles
+// to avoid name collisions every old entry has a `rdstale` suffix
+// if rdstale suffix already exists it uses `{index}-rdstale` to avoid further collisions
+// everything ending with `rdstale` is removed on ack.
+// in case we have key `123` in both old and new result will be: {"123", "123-0-rdstale"}
+// in case old contains {"123", "123-0-rdstale"} and new contains {"123"} result is: {"123", "123-rdstale", "123-0-rdstale"}
+func mergeRoles(zlog zerolog.Logger, old, new *RoleT) (*RoleT, error) {
+	if old == nil {
+		return new, nil
+	}
+	if new == nil {
+		return old, nil
+	}
+
+	oldMap, err := smap.Parse(old.Raw)
+	if err != nil {
+		return nil, err
+	}
+	if oldMap == nil {
+		return new, nil
+	}
+
+	newMap, err := smap.Parse(new.Raw)
+	if err != nil {
+		return nil, err
+	}
+	if newMap == nil {
+		return old, nil
+	}
+
+	destMap := smap.Map{}
+	// copy all from new
+	for k, v := range newMap {
+		destMap[k] = v
+	}
+
+	findNewKey := func(m smap.Map, candidate string) string {
+		if strings.HasSuffix(candidate, "-rdstale") {
+			candidate = strings.TrimSuffix(candidate, "-rdstale")
+			dashIdx := strings.LastIndex(candidate, "-")
+			if dashIdx >= 0 {
+				candidate = candidate[:dashIdx]
+			}
+
+		}
+		// creates new key for permissions
+		// marks it stale so they can be removed later
+		if _, exists := m[candidate]; !exists {
+			return fmt.Sprintf("%s-0-rdstale", candidate)
+		}
+
+		// 1 should be enough, 100 is just to have some space
+		for i := 0; i < 100; i++ {
```

Comment:
You could let just the `for`, no need for this first try, besides the `for` is trying `%s-0-rdstale` again